src/pipeline.rs

```diff
@@ -1,87 +1,115 @@
 use crate::{transforms::FunctionTransform, Event};
-use futures01::{
-    sync::mpsc::{channel, Receiver, SendError, Sender},
-    Async, AsyncSink, Poll, Sink,
-};
-use std::collections::VecDeque;
+use futures::{task::Poll, Sink};
+use std::{collections::VecDeque, fmt, pin::Pin, task::Context};
+use tokio::sync::mpsc;
+
+#[derive(Debug)]
+pub struct ClosedError;
+
+impl fmt::Display for ClosedError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("Pipeline is closed.")
+    }
+}
+
+impl std::error::Error for ClosedError {}
+
+const MAX_ENQUEUED: usize = 1000;
 
 #[derive(Derivative, Clone)]
 #[derivative(Debug)]
 pub struct Pipeline {
-    inner: Sender<Event>,
+    inner: mpsc::Sender<Event>,
     // We really just keep this around in case we need to rebuild.
     #[derivative(Debug = "ignore")]
     inlines: Vec<Box<dyn FunctionTransform>>,
     enqueued: VecDeque<Event>,
 }
 
 impl Pipeline {
-    fn try_flush(&mut self) -> Poll<(), <Self as Sink>::SinkError> {
+    fn try_flush(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), <Self as Sink<Event>>::Error>> {
+        use mpsc::error::TrySendError::*;
+
         while let Some(event) = self.enqueued.pop_front() {
-            if let AsyncSink::NotReady(item) = self.inner.start_send(event)? {
-                self.enqueued.push_front(item);
-                return Ok(Async::NotReady);
+            futures::ready!(self
+                .inner
+                .poll_ready(cx)
+                .map_err(|_: mpsc::error::ClosedError| ClosedError))?;
+
+            match self.inner.try_send(event) {
+                Ok(()) => {
+                    // we good, keep looping
+                }
+                Err(Full(item)) => {
+                    // TODO: can we get here after the poll_ready?
```

Comment:
Nope, `poll_ready` [guarantees](https://docs.rs/futures/0.3.8/futures/channel/mpsc/struct.Sender.html#method.poll_ready) that. So instead of `try_send` we can use `start_send`.

Reply:
Unfortunately, there is no `start_send` on `Sender`, so I think our best option is to keep using `try_send` and consider this branch unreachable. I'll make that change.